spec/app/domain/authentication/authn_iam/Readme.md

```diff
@@ -0,0 +1,165 @@
+    id: authn-iam
+    body:
+    - !policy
+      id: aws
+      body:
+      - !webservice
+
+      - !group authenticatable
+
+      - !permit
+        role: !group authenticatable
+        privilege: [ read, authenticate ]
+        resource: !webservice
+```
+
+### Host Policy
+
+```yaml
+# Applied to the root policy
+- !policy
+  id: aws
+  body:
+  - !policy
+    id: production
+    body:
+    - &hosts
+      - !host 188945769008/my_app_1
+      - !host 188945769008/my-app-2
+      - !host 188945769008/MyApp3
+
+  - !grant
+    members: *hosts
+    role: !group conjur/authn-iam/aws/authenticatable
+```
+
+### How IAM ARNs map to Conjur Roles
+
+The Conjur AWS Authenticator receives an AWS ARN as a result of the authorization request. Below is a sample:
+
+```
+arn:aws:sts::188945769008:assumed-role/conjur-role/i-08241b0e31fe23d20
+```
+
+The interesting bits (from Conjur's perspective) are as follows:
+
+- `188945769008` - AWS account ID
+- `conjur-role` - IAM role given to this workload
+
+All other elements or the AWS ARN are ignored by Conjur.
+
+- The Conjur Role ID must end with `188945769008/conjur-role`
+- The Role can be located in a Conjur policy (ie. can have the prefix: `foo/bar/188945769008/conjur-role`)
+- A Role can be either a User or a Host
+- A Host can be in the Conjur `root` policy
+- A User cannot be in the Conjur `root` policy
+
+## Authentication Workflow
+
+The authenticator works by passing the headers generated by an AWS signing request to the AWS STS's (Security Token Service) `GetCallerIdentity` endpoint.  Conjur uses this signed service to exectute a call to AWS STS (on behalf of the workload we're authenticating).
+
+AWS STS has two options, global (`sts.amazonaws.com`), and regional (`sts.<region>.amazonaws.com`). Some important notes about using the STS endpoint.
+
+- If using the global endpoint (`sts.amazonaws.com`), the request must be signed using the `us-east-1` region. Signing a request using a different region will cause authentication to fail.
+- If using the regional endpoint (preferred by AWS), the request must be signed by signed by that region. For example, if a request to `sts.eu-west-1.amazonaws.com`, it must be signed with the `eu-west-1` region.
+
+### Examples
+
+*All examples are in Ruby.*
+
+*All examples assume the AWS IAM role of the requestor is `my-app-2`*
+
+#### Valid Global Request
+
+```ruby
+require 'aws-sigv4'
+require 'aws-sdk'
+
+signed_aws_headers = Aws::Sigv4::Signer.new(
+  service: 'sts',
+  region: 'us-east-1',
+  credentials_provider: Aws::InstanceProfileCredentials.new
+).sign_request(
+  http_method: 'GET',
+  url: 'https://sts.amazonaws.com/?Action=GetCallerIdentity&Version=2011-06-15'
+).headers
+```
+
+#### Valid Regional Request
+
+Any region can be used as long as the region is the same the target regional STS endpoint.
+
+```ruby
+require 'aws-sigv4'
+require 'aws-sdk'
+
+signed_aws_headers = Aws::Sigv4::Signer.new(
+  service: 'sts',
+  region: 'eu-west-1',
```

Comment:
An EC2 instance can access the meta-data which includes the availability zone the instance is in.  From this information, a user can glean the region the instance is running in.

At some point, it would be really nice to add this signing step into our SDKs and/or create a cleaner solution to generating a signing request from an instance/container/lambda.  It's a burden on the customer, and makes it harder to use the authenticator.